zrml/futarchy/src/pallet_impls.rs

```diff
@@ -0,0 +1,52 @@
+// Copyright 2024 Forecasting Technologies LTD.
+//
+// This file is part of Zeitgeist.
+//
+// Zeitgeist is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the
+// Free Software Foundation, either version 3 of the License, or (at
+// your option) any later version.
+//
+// Zeitgeist is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Zeitgeist. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{types::Proposal, weights::WeightInfoZeitgeist, Config, Event, Pallet};
+use frame_support::{
+    dispatch::RawOrigin,
+    pallet_prelude::Weight,
+    traits::schedule::{v3::Anon, DispatchTime},
+};
+use zeitgeist_primitives::traits::FutarchyOracle;
+
+impl<T: Config> Pallet<T> {
+    /// Evaluates `proposal` using the specified oracle and schedules the contained call if the
+    /// oracle approves.
+    pub(crate) fn maybe_schedule_proposal(proposal: Proposal<T>) -> Weight {
+        let (evaluate_weight, approved) = proposal.oracle.evaluate();
+
+        if approved {
+            let result = T::Scheduler::schedule(
+                DispatchTime::At(proposal.when),
+                None,
+                63,
+                RawOrigin::Root.into(),
+                proposal.call.clone(),
+            );
+
+            if result.is_ok() {
+                Self::deposit_event(Event::<T>::Scheduled { proposal });
+            } else {
+                Self::deposit_event(Event::<T>::UnexpectedSchedulerError);
```

Comment:
This case is unfortunate. This can obviously lead to the following scenario: A proposal is submitted and stored in `Proposals`, but because the scheduler has too many scheduled calls at `proposal.when`, it fails and the whole proposal is removed from the storage, although it has been evaluated to be approved. It might be better to have another extrinsic call that allows approved proposals to be scheduled at the time of the root origins choosing. So essentially this function here could be called `evaluate_proposal` that either emits `Approved` and stores a boolean flag `approved` in the proposal or `Rejected` that removes the proposal. After the successful scheduling from the root origin, the proposal can be removed too.